Python/chapter01/1.1 - Is Unique/miguel_1.1_sol.py

```diff
@@ -0,0 +1,60 @@
+"""
+Python version 3.7.0
+1.1
+Implement an algorithm to determine if a string has all unique characters.
+What if you cannot use additional data structures?
```

Comment:
one thing you should ask (and be able to argue an answer): what about the empty string?

Reply:
Going to push my latest code addressing your review, thank you for the feedback! 👍

Reply:
Looks like the way `all([])` was defined like a universal quantifier from mathematical logic.  It is promised that for all X in S, X is True.  According to https://en.wikipedia.org/wiki/Universal_quantification#The_empty_set
the universal quantifier of an empty set is defined to be true by convention.  Hence why all([]) returns true.  
In mathematical logic, it makes sense that all([]) returns true because we are following the mathematical rules.  
However, in the case of is_unique, I guess I could have the function return true for an empty string, because the problem can be expressed as a universal quantifier (for all c in input_str, there are no two c's that are equal).  But for practical purposes, I think that it's best to let the user know that an empty string was passed in, and they can handle however they wish with a try-catch block. 
What are your thoughts on that?
Source:  https://stackoverflow.com/questions/11979683/why-python-built-in-all-function-returns-true-for-empty-iterables

Reply:
if you are pushing design decisions on the user (e.g. forcing them to catch an exception), the effort required to check for the empty string before they pass it in as an argument is roughly equivalent.

I *personally* prefer to be consistent with the universal qualifier interpretation since it also allows for slightly simpler implementation since you can just ignore the "special case" of the empty string and the code will naturally reach the `return True` (consistency ftw!), but in reality both solutions are correct, as long as you clearly document the precondition.

Reply:
After the user runs into the exception, they can just check for an empty string next time rather than using a try-catch block, so it is forced only if they desire to use the empty string.  So I think it will serve more as a warning, rather than a try catch override.  But then again, it probably depends on the client programmer.

To what point should the input validation be left to the user of the library? (beyond checking for the input type).  Perhaps it should stay flexible.  The user will be guided by the documented precondition, as you have stated.

After digging a little further, I've found that the statement that asserts that all members of an empty set have a certain property is known as the Vacuous Truth.
According to https://en.wikipedia.org/wiki/Vacuous_truth
From the wikipedia link:
> "Outside of mathematics, statements which can be characterized informally as vacuously true can be misleading.  Such statements make reasonable assertions about qualified objects which do not actually exist. For example, a child might tell their parent 'I ate every vegetable on my plate', when there were no vegetables on the child's plate to begin with."

The quote references to statements that are "informally" characterized.  That is not the case here, since we are formally defining the problem.  However, not every programmer (I assume) has a background in formal mathematical logic, so I can see from their perspective why empty-set results as we have been discussing can be misleading.

I understand, though,  that it comes down to personal preference (as you have hinted), and that both approaches are correct.

Reply:
I'm gonna go with your suggestion tho.  Pushing latest code in a few.

Reply:
note that "after the user runs into the exception" should be something you should actively try to design against. this is called the [Principle of Least Astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

being consistent with the rest of the ecosystem, following the surrounding codebase's style, and relying on analogues like the universal qualifier argument are all tools to help this.

> To what point should the input validation be left to the user of the library?

**tl;dr** it depends.

normally the libraries perform validation needed to assert that the preconditions are followed and to give better error diagnostics. for instance, null-checks are rarely performed by libraries since a) you can add type annotations in Java and Python that let the user / compiler know that they shouldn't be passing around null values, and b) the exception thrown by failure to comply with the precondition will be very obvious to the caller. on the other hand, if you have a more complex precondition (say, you are doing an invert matrix routine and the user passes in a [singular matrix](https://en.wikipedia.org/wiki/Invertible_matrix)), unless you do an active validation, the user will be left with a very confusing stack trace when an exception happens in the middle of the function with no obvious reason. if your language and compiler supports helping you with validation in _any_ way (such as type annotations), by all means take advantage of it, since having a compile-time error is much better than having runtime errors.

in the end, everything is a balancing act. you and your team should assign some vague value to the robustness, performance, maintainability, and ergonomics of the codebase. you _can_ write code that meets 100% in all categories (most of the time), but that would consume time dedicated to writing actual business logic. then again, if you _only_ focus on features business logic and neglect everything else, you most likely will end up with an unmaintainable spaghetti code mess that's awful to use and unexplainably slow.

Reply:
Noted. 👍 
Thank you, once again, for your feedback.  This was all incredibly helpful.